src/tool_operate.c

```diff
@@ -486,7 +488,51 @@ static CURLcode retrycheck(struct OperationConfig *config,
 
     per->retry_remaining--;
 
-    if(outs->bytes && outs->filename && outs->stream) {
+    /* Skip truncation of outfile if auto-resume is enabled for download and
+       the partially received data is good. Currently this is only for HTTP
+       GET requests in limited circumstances. */
+    if(config->use_resume && config->resume_from_current &&
+       config->resume_from >= 0 && outs->init == config->resume_from &&
+       (config->httpreq == TOOL_HTTPREQ_UNSPEC ||
+        config->httpreq == TOOL_HTTPREQ_GET) &&
+       (!config->customrequest || !strcmp(config->customrequest, "GET")) &&
+       !config->use_ascii && !per->uploadfile &&
+       result != CURLE_WRITE_ERROR && result != CURLE_RANGE_ERROR &&
+       outs->bytes > 0 && outs->filename && outs->stream) {
+      long response = 0;
+      const char *method = NULL, *scheme = NULL;
+
+      curl_easy_getinfo(curl, CURLINFO_EFFECTIVE_METHOD, &method);
+      curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response);
+      curl_easy_getinfo(curl, CURLINFO_SCHEME, &scheme);
+      scheme = proto_token(scheme);
+
+      if((scheme == proto_http || scheme == proto_https) &&
+         method && !strcmp(method, "GET") &&
+         response == (config->resume_from ? 206 : 200)) {
```

Comment:
I would expect the server to also send `Accept-Ranges: bytes` to signal it accepts range requests (using byte ranges).

Reply:
> I would expect the server to also send Accept-Ranges: bytes to signal it accepts range requests (using byte ranges).

Do most servers do this in practice? I'll add a check for it when response code is 200 (and not 206).